plugins/processors/snmp_lookup/lookup.go

```diff
@@ -0,0 +1,338 @@
+	minRetry               = 5 * time.Minute
+	orderedQueueSize       = 10_000
+)
+
+func (*Lookup) SampleConfig() string {
+	return sampleConfig
+}
+
+func (l *Lookup) Init() (err error) {
+	l.sigs = make(signalMap)
+	l.getConnectionFunc = l.getConnection
+
+	if _, err = snmp.NewWrapper(l.ClientConfig); err != nil {
+		return fmt.Errorf("parsing SNMP client config: %w", err)
+	}
+
+	switch l.Translator {
+	case "", "gosmi":
+		if l.translator, err = si.NewGosmiTranslator(l.Path, l.Log); err != nil {
+			return fmt.Errorf("loading translator: %w", err)
+		}
+	default:
+		return fmt.Errorf("invalid translator %q", l.Translator)
+	}
+
+	// Initialize the table
+	l.table.Name = "lookup"
+	l.table.IndexAsTag = true
+	l.table.Fields = make([]si.Field, len(l.Tags))
+	for i, tag := range l.Tags {
+		tag.IsTag = true
+		l.table.Fields[i] = tag
+	}
+
+	return l.table.Init(l.translator)
+}
+
+func (l *Lookup) Start(acc telegraf.Accumulator) error {
+	l.cache = expirable.NewLRU[string, tagMap](l.CacheSize, nil, time.Duration(l.CacheTTL))
+	if l.Ordered {
+		l.parallel = parallel.NewOrdered(acc, l.addAsync, orderedQueueSize, l.ParallelLookups)
+	} else {
+		l.parallel = parallel.NewUnordered(acc, l.addAsync, l.ParallelLookups)
+	}
+	return nil
+}
+
+func (l *Lookup) Add(metric telegraf.Metric, _ telegraf.Accumulator) error {
+	l.parallel.Enqueue(metric)
+	return nil
+}
+
+func (l *Lookup) addAsync(metric telegraf.Metric) []telegraf.Metric {
+	if !metric.HasTag(l.AgentTag) {
+		l.Log.Warn("Agent tag missing")
+		return []telegraf.Metric{metric}
+	}
+
+	index, ok := metric.GetTag(l.IndexTag)
+	if !ok {
+		l.Log.Warn("Index tag missing")
+		return []telegraf.Metric{metric}
+	}
+
+	gs, err := l.getConnectionFunc(metric)
+	if err != nil {
+		l.Log.Errorf("Could not prepare connection: %v", err)
+		return []telegraf.Metric{metric}
+	}
+
+	// Prepare cache
+	if err := l.prepareCache(gs, index); err != nil {
+		l.Log.Warnf("Could not prepare cache for %q: %v", gs.Host(), err)
+		return []telegraf.Metric{metric}
+	}
+
+	// Load from cache
+	tagMap, inCache := l.cache.Get(gs.Host())
+	tags, indexExists := tagMap.rows[index]
+	if inCache && indexExists {
+		for key, value := range tags {
+			metric.AddTag(key, value)
+		}
+	} else {
+		l.Log.Warnf("Could not find index %q on agent %q", index, gs.Host())
+	}
+
+	return []telegraf.Metric{metric}
+}
+
+// prepareCache prepares the cache if needed (index does not exist yet, or agent not yet cached)
+func (l *Lookup) prepareCache(gs snmpConnection, index string) error {
+	agent := gs.Host()
+
+	// Check cache
+	l.lock.Lock()
+	tagMap, inCache := l.cache.Peek(agent)
+	_, indexExists := tagMap.rows[index]
+
+	// Cache miss or non existing index and not recently refreshed the table
+	if !inCache || (!indexExists && time.Since(tagMap.created) > minRetry) {
+		// Check if another process is alreay loading the table, wait until done if so
+		if done, busy := l.sigs[agent]; busy {
+			l.lock.Unlock()
+			<-done
```

Comment:
I'm still seeing major issues here. The most important one is that if you set `l.ParallelLookups` to say `4` and then you get in 10 different metrics (from 4 different agents) but the first 4 are from the same `agent` you will block the entire thing and you are running only one `agent` query instead of the 4 possible ones!

Reply:
That's indeed correct. It is probably also the reason why default parallel lookups is set to 100.

I don't see an easy way to workaround this. And also, is it really an issue? All I see is it would take a bit more to build the cache and return the metrics..

Reply:
>  It is probably also the reason why default parallel lookups is set to 100.

Whereas on the other hand if the metrics are ordered correctly and contain 100 agents you send out 100 queries in parallel. Effectively, this mean the parameter is worthless...

> I don't see an easy way to workaround this.
I proposed one, but you didn't like it... ;-)

Reply:
> I proposed one, but you didn't like it... ;-)

Okay, if we're going back that route, please finally give feedback in the main thread on my 2 remarks I expressed here https://github.com/influxdata/telegraf/pull/14223#issuecomment-1851842807. IMO they are quite valid..

> Whereas on the other hand if the metrics are ordered correctly and contain 100 agents you send out 100 queries in parallel. Effectively, this mean the parameter is worthless...

I'm not following, why would the parameter be worthless? Do you mean the number of parallel threads should be limitless?

Reply:
I also have a new idea to not use the `parallel.Enqueue` when the respective index on the agent is already cached. So only enqueue when the agent or index isn't in the cache, otherwise directly apply the tags to the metric..